File: OrangeJuice.Server.Core/Security/X509Certificate2Factory.cs
using System;
using System.Linq;
using System.Security.Cryptography;
using System.Security.Cryptography.X509Certificates;

using OrangeJuice.Server.Configuration;

namespace OrangeJuice.Server.Security
{
	public sealed class X509Certificate2Factory : Factory.IFactory<X509Certificate2>
	{
		private readonly GoogleAuthOptions _authOptions;
		private readonly IEnvironmentProvider _environmentProvider;

		public X509Certificate2Factory(GoogleAuthOptions authOptions, IEnvironmentProvider environmentProvider)
		{
			_authOptions = authOptions;
			_environmentProvider = environmentProvider;
		}

		public X509Certificate2 Create()
		{
			switch (_environmentProvider.GetCurrentEnvironment())
			{
				case EnvironmentName.Production:
					return Create(_authOptions.CertificateThumbprint);
				default:
					return Create(_authOptions.CertificateKey, _authOptions.CertificateSecret);
			}
		}

		private static X509Certificate2 Create(string key, string secret)
		{
			return new X509Certificate2(Convert.FromBase64String(key), secret);
		}

		private static X509Certificate2 Create(string thumbprint)
		{
			var certStore = new X509Store(StoreName.My, StoreLocation.CurrentUser);
			certStore.Open(OpenFlags.ReadOnly);

			var certCollection = certStore.Certificates.Find(X509FindType.FindByThumbprint, thumbprint, false);
			var cert = certCollection.Cast<X509Certificate2>().FirstOrDefault();

			certStore.Close();

			if (cert == null)
			{
				throw new CryptographicException(String.Format("Certificate with thumbprint '{0}' was not found in store '{1}'", thumbprint, certStore.Name));
			}
			return cert;
		}
	}
}

File: OrangeJuice.Server.Core.FSharp/Security/X509Certificate2Factory.fs
namespace OrangeJuice.Server.FSharp.Security

open System
open System.Security.Cryptography
open System.Security.Cryptography.X509Certificates

open OrangeJuice.Server.Configuration
open OrangeJuice.Server.Security

type X509Certificate2Factory(authOptions : GoogleAuthOptions, environmentProvider : IEnvironmentProvider) =
    interface Factory.IFactory<X509Certificate2> with
        member this.Create() : X509Certificate2 =
            match environmentProvider.GetCurrentEnvironment() with
            | EnvironmentName.Production -> X509Certificate2Factory.Create(authOptions.CertificateThumbprint)
            | _ -> X509Certificate2Factory.Create(authOptions.CertificateKey, authOptions.CertificateSecret)

    static member Create(key : string, secret : string) : X509Certificate2 =
        new X509Certificate2(Convert.FromBase64String(key), secret)

    static member Create(thumbprint : string) : X509Certificate2 =
        let certStore = new X509Store(StoreName.My, StoreLocation.CurrentUser)
        certStore.Open(OpenFlags.ReadOnly) |> ignore

        let certCollection = certStore.Certificates.Find(X509FindType.FindByThumbprint, thumbprint, false)
        let cert = certCollection
                   |> Seq.cast<X509Certificate2>
                   |> Seq.head

        certStore.Close() |> ignore

        if cert <> null then cert
        else raise <| new CryptographicException(sprintf "Certificate with thumbprint '%s' was not found in store '%s'" thumbprint certStore.Name)
